### DISCUSSION TITLE
SW360 Frontend

### DISCUSSION THREAD

**Discussion by heliocastro** [eclipse-sw360/sw360#1784] [Ideas]
As possible replacement from [Liferay frontend](https://liferay.dev/), and to simplify SW360 deployment, there are initiatives to replace to a external modern Frontend work

Previous attempt using [Angular JS](https://angularjs.org/) was done by [Bosch.io in this referenced branch](https://github.com/eclipse/sw360/tree/feature/client-angular-ui) but was not moved ahead and became outdated against current SW360 code.  

Usage of this branch will be same as start from scratch complete new, which open up to  possibilities, but for effect to reduce the choice amount, previous research on frameworks lead to 3 possible candidates referred below.

This discussion already start with possible winner horse, [Svelte Kit](https://kit.svelte.dev/), sine there are a full functional proof of concept done on the beginning of SW360 16 cycle release, made here [sw360-frontend](https://github.com/heliocastro/sw360-frontend)

One of the main factors considered is that current SW360 developers aren't web or frontend developers originally, so any framework **MUST** have an easy learning curve and have documentations already

Second, and important: **JAVASCRIPT  is a different mind set**. For SW360 developers reality, choose framework by popularity is not an option, since is necessary an entire learning process, including how a javascript environment runs, how an web application and reactive elements work. 

Some other details that are basic requirements:
* Need support CSS integrations ( like [Tailwind CSS](https://tailwindcss.com/) )
* Need support custom widgets
* Need to be easy to setup a dev environment to not intimidate new developers

The possible candidates, set on order of preference:

1 - [Svelte Kit](https://kit.svelte.dev/)
     Svelte enter in the top of the list as:
      * huge growing in the usage on last 2 years
      * Modern component oriented framework
      * Easy curve adoption
      * Most developed framework today
      * Excelent documentation

2 - [React JS](https://reactjs.org/)
     React is an obvious secondary choice, just as second due the DOM model making the learning curve hard, and slow development. It has though
     * Large base adopted 
     * Plenty of documentation
     * Easy to find resources
    
3 - [Vue JS](https://v2.vuejs.org/)
     Vue is the third, as the middle ground between React and Svelte, and with strong community but lack of the easy learning curve

Comparisons  found to a general idea:
* https://dev.to/heyjtk/vue-vs-svelte-vs-react-3l31
* https://joshcollinsworth.com/blog/introducing-svelte-comparing-with-react-vue
* https://www.openxcell.com/blog/svelte-vs-react/

**Comment by ag4ums** [comment]
@heliocastro ,

[Svelte Kit](https://kit.svelte.dev/) is good but its very young in the market... and maturing... so now it will be very early to adapt svelte for a enterprise application like sw360.

Angular and React its difficult to choose, as `Angular is Framework base solution`, so Angular will be more preferred solution for sw360 UI.
and for distributed project like sw360 it will be more beneficial to use a framework based solution that provide rigidity in development and also don't mixes code like html and java script.

**Comment by KoukiHama** [comment]
Sorry for the late reply.

As I mentioned in Slack, I would recommend React.

Svelte is still in its infancy.

Angular on the other hand is complex and expensive to learn. So, React is the best choice.

It is also a popular and friendly framework for those who want to develop many SW360s.
I summarize React good points.

* Virtual DOM: React's virtual DOM (Document Object Model) allows for efficient updates and rendering of components, resulting in better performance.

* Popularity: React is widely used and has a large and active community, which makes it easier to find help and resources.

* Flexibility: React can be used with a variety of different programming languages and tools, making it more flexible than some other frameworks.

And you can see how React is popular by google trends.
https://trends.google.co.jp/trends/explore?q=React,Angular,Svelte,Vue

and  npm trends is also helpful our discussion.

https://npmtrends.com/angular-vs-react-vs-svelte-vs-vue

**Reply by KoukiHama** [reply]
We have started to make poc of react frontend there
https://github.com/toshiba/sw360/commits/dev/react-ui

**Reply by ag4ums** [reply]
poc is good but we need to know what use-case/s, you are started working on, so that we can collaborate, without overlap/duplicate work, and also please raise a PR for the same.

**Reply by KoukiHama** [reply]
Now, we are simply using React to mimic the current GUI. 
Next we plan to create a simple/easy login function.
This is also the purpose of studying React.

**Reply by KoukiHama** [reply]
And as you point out, we need to consider use cases to avoid duplication of work.
We haven't done anything meaningful yet in the above branch. Let's discuss how we can collaborate in the future.

**Reply by tienlee** [reply]
Hello, we have started to make POC of react frontend and created a simple/easy login function at https://github.com/toshiba/sw360/tree/dev/react-ui/sw360-frontend

Could you please try this function and let me know if the folder structure should be changed or not?

In addition, do you know Redux (https://redux.js.org/)? Should we use Redux?

**Reply by KoukiHama** [reply]
@ag4ums 
We can raise this PR (https://github.com/eclipse/sw360/compare/main...toshiba:sw360:dev/react-ui) with "**do not merge**" label, 
Is this way more convenient for use case discussion and collaboration?

Also, as telco talked about, if you already have the login function source code, we may be able to merge it because what we have now is very simple.

**Reply by KoukiHama** [reply]
https://github.com/eclipse/sw360/pull/1839
This is PR for discussion

**Comment by heliocastro** [comment]
Closing this discussion, We're alive and well now :-)
